### PR TITLE
admin snapshots frontend

### PR DESCRIPTION
**Pull request by emilia-friedberg**
## WHAT
This PR constitutes the bulk of the work for the admin snapshots frontend, utilizing all the backend functionality available as of 06/02. It's primarily React and CSS code, with some minor adjustments to the backend where necessary to get it working in testing.

Please note: the number of files here are very large, but a lot of it is just snapshot updates (I centralized some shared mock logic so I didn't have to repeat it in a bunch of tests).

## WHY
We want this exciting feature to be functional through the UX!

## HOW
Build out the frontend to feature completion with placeholder and "Coming soon" states where there isn't yet backend support and full functionality where there is. The implementation is fully mobile responsive. Connection to the backend for snapshot ranking components will happen in a future PR, as this one is already big enough.

### Screenshots
Screen Shot 2023-06-02 at 2 04 01 PM
Screen Shot 2023-06-02 at 2 04 15 PM
Screen Shot 2023-06-02 at 2 04 41 PM
Screen Shot 2023-06-02 at 2 07 38 PM
Screen Shot 2023-06-02 at 2 08 38 PM
Screen Shot 2023-06-02 at 2 09 06 PM
Screen Shot 2023-06-02 at 2 09 35 PM
Screen Shot 2023-06-02 at 2 10 02 PM

### Notion Card Links
[(Please provide links to any relevant Notion card(s) relevant to this PR.)
](https://www.notion.so/quill/Admin-Report-Front-End-UI-Usage-Snapshot-Data-Export-07c83420bd774e72aa8657748fc997c1?pvs=4)

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? | YES
Have you deployed to Staging? | Not yet - deploying now!
Self-Review: Have you done an initial self-review of the code below on Github? | YES
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | YES

### CODE REVIEW

File: services/QuillLMS/client/app/bundles/PremiumHub/components/usage_snapshots/shared.ts [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+const PRACTICE = 'Practice'
+const CLASSROOMS = 'Classrooms'
+const SCHOOLS = 'Schools'
+
+// custom date
+export const CUSTOM = 'custom'
+
+export const TAB_NAMES = [ALL, HIGHLIGHTS, USERS, PRACTICE, CLASSROOMS, SCHOOLS]
+
+export const snapshotSections = [
```
Comment (anathomical) [inline services/QuillLMS/client/app/bundles/PremiumHub/components/usage_snapshots/shared.ts:48]: I will say that I really like the way you've designed this to be so configurable as a big object.  Should make it so easy to update and changes as requireed.

File: services/QuillLMS/client/app/bundles/PremiumHub/components/usage_snapshots/shared.ts [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+  }
+
+  for (let key of obj1Keys) {
+    if (obj1[key] !== obj2[key]) {
+      return false;
+    }
+  }
+
+  return true;
+}
```
Comment (anathomical) [inline services/QuillLMS/client/app/bundles/PremiumHub/components/usage_snapshots/shared.ts]: It kind of feels that these are generalizable enough that we might want them shared at an even higher level of the directory hierarchy.  Not a problem leaving them here, but just a thought.

File: services/QuillLMS/client/app/bundles/PremiumHub/components/usage_snapshots/snapshotCount.tsx [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+      const queryKeysAreEqual = message.query === queryKey
+      const timeframesAreEqual = message.timeframe === selectedTimeframe
+      const schoolIdsAreEqual = _.isEqual(message.school_ids, selectedSchoolIds.map(id => String(id)))
+      const gradesAreEqual =  _.isEqual(message.grades, selectedGrades.map(grade => String(grade)))
+
+      if (queryKeysAreEqual && timeframesAreEqual && schoolIdsAreEqual && gradesAreEqual) {
+        getData()
+      }
+    });
+  };
```
Comment (anathomical) [inline services/QuillLMS/client/app/bundles/PremiumHub/components/usage_snapshots/snapshotCount.tsx]: I really like the check to make sure that the Pusher filters match the current filters before re-fetching data.  Very nice.

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/UsageSnapshotsContainer.tsx [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+      const timeframeOptions = filterData.timeframes.map(tf => ({ ...tf, label: tf.name }))
+      const gradeOptions = filterData.grades.map(grade => ({ ...grade, label: grade.name }))
+      const schoolOptions = filterData.schools.map(school => ({ ...school, label: school.name, value: school.id }))
+
+      const timeframe = defaultTimeframe(timeframeOptions)
+
+      setAllGrades(gradeOptions)
+      setAllTimeframes(timeframeOptions)
+      setAllSchools(schoolOptions)
+      setSelectedGrades(gradeOptions)
```
Comment (anathomical) [inline services/QuillLMS/client/app/bundles/PremiumHub/containers/UsageSnapshotsContainer.tsx:108]: Per our discussion on Slack, I think we want our default filter value for `selectedGrades` to be `[]` in order to avoid filtering out classrooms that don't have an assigned grade.

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/UsageSnapshotsContainer.tsx [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+      setSelectedSchools(schoolOptions)
+      setLastUsedTimeframe(timeframe)
+      setSelectedTimeframe(timeframe)
+      setLoadingFilters(false)
+    })
+  }
+
+  function clearFilters() {
+    setSelectedGrades(allGrades)
```
Comment (anathomical) [inline services/QuillLMS/client/app/bundles/PremiumHub/containers/UsageSnapshotsContainer.tsx:117]: Echoing my above comment, I think we want to reset `selectedGrades` to `[]` when we clear filters so that we avoid filtering out classrooms without grade levels selected.

**Review by anathomical — APPROVED** [review]
One comment on default `selectedGrades` values, which we discussed on Slack, but overall this looks really great.  I love the way you've organized config into an object for quick and easy tweaks.

File: services/QuillLMS/client/app/bundles/PremiumHub/components/usage_snapshots/shared.ts (hunk 1 above)
Comment (eadams17) [reply services/QuillLMS/client/app/bundles/PremiumHub/components/usage_snapshots/shared.ts:48]: agreed!

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/UsageSnapshotsContainer.tsx [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+  }, [])
+
+  React.useEffect(() => {
+    if (loadingFilters) { return }
+
+    const newValueForHasAdjustedFiltersFromDefault = !unorderedArraysAreEqual(selectedSchools, allSchools) || !unorderedArraysAreEqual(selectedGrades, allGrades) || !unorderedArraysAreEqual(selectedTimeframe, defaultTimeframe(allTimeframes))
+
+    setHasAdjustedFiltersFromDefault(newValueForHasAdjustedFiltersFromDefault)
+
+    const newValueForHasAdjustedFiltersSinceLastSubmission = (!unorderedArraysAreEqual(selectedSchools, lastSubmittedSchools) || !unorderedArraysAreEqual(selectedGrades, lastSubmittedGrades) || !unorderedArraysAreEqual(selectedTimeframe, lastSubmittedTimeframe) || customStartDate !== lastSubmittedCustomStartDate || customEndDate !== lastSubmittedCustomEndDate)
```
Comment (eadams17) [inline services/QuillLMS/client/app/bundles/PremiumHub/containers/UsageSnapshotsContainer.tsx]: this is a small take or leave nit, but this line is pretty long and was a bit hard to parse at first-- so maybe it might be more readable to separate out like this?

```
const arraysUnequal = (!unorderedArraysAreEqual(selectedSchools, lastSubmittedSchools) || !unorderedArraysAreEqual(selectedGrades, lastSubmittedGrades) || !unorderedArraysAreEqual(selectedTimeframe, lastSubmittedTimeframe)

const datesDoNotMatch = customStartDate !== lastSubmittedCustomStartDate || customEndDate !== lastSubmittedCustomEndDate

const newValueForHasAdjustedFiltersSinceLastSubmission = arraysUnequal || datesDoNotMatch
```

**Review by eadams17 — APPROVED** [review]
Thank you for putting all of this together! Everything is very well organized

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/UsageSnapshotsContainer.tsx (hunk 4 above)
Comment (emilia-friedberg) [reply services/QuillLMS/client/app/bundles/PremiumHub/containers/UsageSnapshotsContainer.tsx:108]: I took a slightly different approach here (`selectedGrades` here remains the same, but when we format it to pass it into the `SnapshotSection` components where the requests actually get made, we pass an empty array if it's full). Otherwise we'd have to do employ much more complicated logic to pass in and out the correct values from the `DropdownInput` component. The end result is the same - the backend gets an empty array when all grades are selected.

Comment (anathomical) [reply services/QuillLMS/client/app/bundles/PremiumHub/containers/UsageSnapshotsContainer.tsx:108]: That seems totally reasonable.
